src/config.rs

```diff
@@ -27,32 +27,35 @@ pub struct ServerConfig {
 #[derive(Deserialize, Debug)]
 pub struct RpcMethod {
     pub method: String,
-    pub params: Vec<MethodParam>,
+
+    #[serde(default)]
+    pub params: Vec<ParamType>,
 
     #[serde(default)]
     pub cache: usize,
 }
 
 impl RpcMethod {
-    pub fn inject_block_num(&self) -> Option<usize> {
+    pub fn at(&self) -> Option<(usize, &ParamType)> {
         self.params
             .iter()
-            .position(|p| p.inject == Some(true) && p.ty == "BlockNumber")
-    }
-
-    pub fn inject_block_hash(&self) -> Option<usize> {
-        self.params
-            .iter()
-            .position(|p| p.inject == Some(true) && p.ty == "BlockHash")
+            .enumerate()
+            .rfind(|(_index, param)| matches!(param, ParamType::At(_)))
     }
 }
 
-#[derive(Deserialize, Debug, Eq, PartialEq)]
-pub struct MethodParam {
-    pub name: String,
-    pub ty: String,
-    pub is_optional: Option<bool>,
-    pub inject: Option<bool>,
+#[derive(Clone, Deserialize, Debug)]
+pub enum ParamType {
+    BlockHash,
+    BlockNumber,
+    #[serde(rename = "vec")]
+    Vec(Box<ParamType>),
+    #[serde(rename = "optional")]
+    Optional(Box<ParamType>),
+    #[serde(rename = "at")]
+    At(Box<ParamType>),
```

Comment:
will be good to add docstring as the `At` type implication (optional, inject, block hash or number) is not self-explaining from its name